content/into.md

```diff
@@ -0,0 +1,328 @@
+
+Let's first take a look the `Expr` example, which uses into-as-a-constructor. Could it be rewritten to use into-as-a-modifier?
+This would mean we have to add `into` to the whole `Expr` enum. Adding it to just `Const` is not enough, since `Add` and `Neg` take `Expr` arguments, not `Const` arguments.
+
+But we might not always have permission to change the `Expr` enum. For instance, `Expr` could be defined in a lower level library without implicit conversions, but later we want to make `Expr` construction convenient by eliding `Const` wrappers in some higher-level library or application. With `into` constructors, this is easy: Define the implicit conversion and facade methods that construct `Expr` trees while taking `into[Expr]` parameters.
+With `into` modifiers there is no way to achieve the same.
+
+A possibly more important objection is that even if we could add the `into` modifier to `Expr`, it would be bad style to do so! We want to allow for implicit conversion in the very specific case where we build an `Expr` tree using the `Add` and `Neg` constructors. Our applications could have lots of other methods that take `Expr` trees, for instance to analyze them or evaluate them.
+We probably do not want to allow implicit conversions for the arguments of all these other methods. The `into` modifier is too unspecific to distinguish the good use case from the problematic ones.
+
+On the other hand, there are also situations where into-as-a-modifier is the practical choice. To see this, consider again the `Modifier` use case in Laminar.
+We could avoid the `into` modifier by wrapping all `Modifier` parameters
+with the `into` constructor. This would be a lot more work than adding just the single `into` modifier. Worse, functions taking `Modifier` parameters are found both in the Laminar framework code and in many applications using it. The framework and the applications would have to be upgraded in lockstep. When Laminar upgrades to Scala 3 implicits, all applications would have to be rewritten, which would make such a migration very cumbersome.
+
+One can try to mitigate the effort by playing with type aliases. For instance, a hypothetical future Laminar using Scala 3 conversions could rename the
+trait `Modifier` to `ModifierTrait` and define an alias
+```scala
+type Modifier = into[ModifierTrait]
+```
+Then the source code of applications would not have to change (unless these applications define classes directly extending `Modifier`). But that future Laminar would not be binary compatible with the current one, since the name
+of the original `Modifier` trait has changed. In summary, upgrading Laminar to use Scala 3 conversions could keep either source compatibility or binary compatibility but not both at the same time.
+
+
+## Syntax Changes
+
+```
+LocalModifier     ::=  ...  |  ‘into’
+```
+
+`into` is a soft modifier. It is only allowed on classes, traits, and opaque type aliases.
```

Comment:
It was only now that I realized that enums are not included here. Was this intentional? @odersky

Reply:
I tried this snippet with the current nightly compiler (3.7.3-RC1-bin-20250619-b1da8fc-NIGHTLY):
```scala
import scala.language.experimental.into

into enum Foo:
  case Foo1
  case Foo2
```
and compiling resulted in the error:
```
[error] Only access modifiers are allowed on enum definitions
[error] into enum Foo:
[error]           ^^^
```
so it looks like this doesn't work and it wasn't clear for me if that's because of the spec or the current implementation

Reply:
Yes. On second thought I think we should leave the spec and implementation as is. The rule that enums take only access modifiers is valuable, and the ability to add `into` as a modifier to enums is a bit dubious since `into` is primarily intended for migration from Scala 2 codebases. But of course there we no enums in Scala 2.